### ISSUE TITLE
[core] Actors don't get cleaned up on pytest failures with scoped ray session

### ISSUE THREAD

**Issue by krfricke** [ray-project/ray#31738]
### What happened + What you expected to happen

When using pytest to run unit tests, we re-use Ray cluster across tests to avoid cluster start/teardown overhead.

However, when a test fails, it's scheduling state is not properly cleaned up. For instance, actors are still running and take up resources. Subsequent tests can then fail or hang due to previous state.

See the reproduction script for a simple example. The second test hangs forever as the first tests actor still occupies resources, even though the actor object went out of scope and should be garbage collected.
The same test works if the `raise RuntimeError` line is commented out.

### Versions / Dependencies

master

### Reproduction script

```
import pytest
import ray


@pytest.fixture(scope="module")
def ray_start_4_cpus():
    address_info = ray.init(num_cpus=4)
    yield address_info
    ray.shutdown()


@ray.remote(num_cpus=4)
class Actor:
    def task(self):
        return True


@pytest.mark.parametrize("failing", [True, False])
def test_failing(ray_start_4_cpus, failing):
    actor = Actor.remote()
    print(ray.get(actor.task.remote()))
    if failing:
        print("Failing")
        raise RuntimeError  # comment out to make both tests pass
    print("Done")
```

```
pytest -s file.py
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by rkooo567** [comment]
I ran this locally, and it seems to raise an exception (which means things didn't hang)?

**Comment by rkooo567** [comment]
I think it is just the test issue. I added 

```
@pytest.fixture(scope="module")
def ray_start_4_cpus():
    address_info = ray.init(num_cpus=4)
    print("init")
    yield address_info
    print("SANG-TODO shutdown")
    ray.shutdown()
```
and then SANG-TODO shutdown is not printed until both tasks are completed

**Comment by rkooo567** [comment]
So I think it is just misuse of fixture (it doesn't seem to be applied when it is retried with a parameter)

**Comment by rkooo567** [comment]
Hmm actually are you saying when we run the second loop, the actor should be removed because the ref should go out of scope?

**Comment by rkooo567** [comment]
So, I did a bit of investigation for this. 

TL;DR I recommend you to run ray.init() and ray.shutdown in between each test although you use the same cluster

Observation

1. I found the actor from the previous run eventually is GC'ed after some time.
2. I found if I use gc.collect() in the beginning of the test run, it finishes immediately.

So I think it is just some Python GCs the previous pytest's actor reference a little late
